### ISSUE TITLE
twtxt pod instances use excessive traffic

### ISSUE THREAD

**Issue by hecanjog** [jointwt/twtxt#427]
The vast majority of outgoing traffic on my site is serving my twtxt.txt file to pod instances. It's not crippling my server or anything, but it seems excessive. Requests are coming in from pods every few seconds.

I'm not at the point of blocking twtxt pod HTTP traffic, but would you please consider reducing the default update interval to something more reasonable? Maybe every hour?

I'd be happy to add some special indication in my twtxt file to request less frequent access, too.

**Comment by prologic** [comment]
Very few seconds seems unlikely, unless you have lots of followers? The default configuration is 5m. However it also respects caching and `Last-Modified` checks. How are you serving up your feed?

**Comment by prologic** [comment]
This isn't the first time I've seen/heard of this kind of complain, so I sympathize, but if I'm to be honest, if the `If-Last-Modified-Since` check is working correctly all you should see is an small/empty `GET` which will not generate much traffic at all.

That being said, rather then making the "polling interval" longer, I'd rather I spend time/energy writing a smarter poller that does some kind of backoff for feeds that don't see frequent updates. So (_for example_) if you only post once a week or so, pods may only check back on your feed say once every few hours. If you post every day or multiple times a day it would check back every hour or 1/2hr.

**Comment by prologic** [comment]
cc @aaadonai @der-lyse @jlj77 @JonLundy for comments

**Comment by aaadonai** [comment]
I like the idea of a smarter polling. Also the idea of the twtxt owner providing the interval the instances should request the twtxt seems like a good approach.
@prologic How would you approach the smart polling? I mean in terms of calculating the right interval for each twtxt feed. Do you think an average of the last x posts intervals would work? Or should it be an average of the last x days of posts?

**Comment by prologic** [comment]
> I like the idea of a smarter polling. Also the idea of the twtxt owner providing the interval the instances should request the twtxt seems like a good approach.

Yes as a first step we can make it configurable. I'll see if I can add this soon.

> @prologic How would you approach the smart polling? I mean in terms of calculating the right interval for each twtxt feed. Do you think an average of the last x posts intervals would work? Or should it be an average of the last x days of posts?

To be honest I'm not sure. I haven't given it too much thought. But _ think_ you'd want to end up with something reasonable so frequent posters you see in reasonable time and less frequent posters you don't hammer their web server(s) that host their feed.

**Comment by der-lyse** [comment]
Specifying a poll interval for the twtd operator in the config seems like a very good and easy solution. So let's go with this at first.

I also like the idea of a weighted interval calculation, but getting it right is tricky. So that would be the second thing to do.

Maybe we invent another extension that lets twtxt feed owners specify a crawl interval, that could be honored by twtd and other clients. I reckon, this would be the last thing to work on, if that. [The specification on Syndication](https://web.resource.org/rss/1.0/modules/syndication/) might inspire us.

**Comment by hecanjog** [comment]
I just wanted to say I appreciate you all giving this some thought!

One additional idea to consider could be just to disable polling at some interval after the user has stopped interacting with the site. You'd have a little warmup time at the first interaction I suppose, but this would probably reduce traffic by a lot. (I doubt even heavy users need to interact with their pods more than a few hours a day? Though if there are background notifications and such that use polling this idea is moot!)

In other words, if there hasn't been an interaction from the user in say 5 minutes, then stop polling. On the next user interaction, start polling again...

Just a thought. Thanks again for considering this!

**Comment by hecanjog** [comment]
> How are you serving up your feed?

It's a static text file served by apache. FWIW I did spend some time looking into rate limiting with the server. At the moment it's more trouble than it is worth. :)

**Comment by der-lyse** [comment]
> In other words, if there hasn't been an interaction from the user in say 5 minutes, then stop polling. On the next user interaction, start polling again...

Oh, that's an interesting idea, I like it a lot!

However, I reckon this could be quite a bit difficult to accomplish, since the polling is decoupled from everything else at the moment. So the scraping job is similar to a cron job running alone in the background and doing its thing without caring or being able to see if there are any active user sessions or not. I'm not saying it cannot be done, just that it's not that easy as it may sound. Disclaimer: I didn't have a look at the code, so I could be completely wrong – wouldn't be the first time. ;-)

**Comment by prologic** [comment]
So after reading all the comments, there are some good points here. Here's what I'm thinking so far:

- Adding a configurable pol interval for pod owner/operators makes sense. This should have been done long ago.
  - However doing so doesn't necessarily address @hecanjog or anyone else's concerns about hitting feeds "too hard" (_for various definitions_), mostly because I cannot control (_unless I close registrations_) how large my own pod is and how many users it has and their interactions.
 
That being said:

- We should flesh out the 2nd option IHMO. Let's continue the discussion while I cut a PR to make pod polling intervals configurable.

The other option that was discussed won't really work very well @der-lyse is right, the "cache" that is shared by all users is really a "global cache" and the job that periodically updates it is effectively a background cron job (_not the only background jobs that run on pods, but one of the main ones_) -- Therefore it is impractical from a design point-of-view to change this.

So... How should we come up with a decent algorithm that intelligently backs off polling one feed but not another and strikes a god balance?

The final option of adding to the spec let's table that for a later date. I _think_ we've made [significant additions to the spec](https://dev.twtxt.net) so far and adding another one that seems rather specific to multi-user setups like this probably won't be all that useful in general or widely adopted. As such only Twt Subjects / Twt Hashes and Avatars are adopted somewhat by more newere/modern clients/servers.
